llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -7325,6 +7333,18 @@ bool SimplifyCFGOpt::simplifyCondBranch(BranchInst *BI, IRBuilder<> &Builder) {
           if (mergeConditionalStores(PBI, BI, DTU, DL, TTI))
             return requestResimplify();
 
+  // Check if the condition is an equality between two constants. This can form
+  // due to other CFGSimplify steps, and may prevent further simplification if
+  // we don't deal with it here.
+  if (auto ICmp = dyn_cast<ICmpInst>(BI->getCondition()))
+    if (ICmp->getPredicate() == CmpInst::ICMP_EQ)
+      if (auto *LHS = dyn_cast<ConstantInt>(ICmp->getOperand(0)))
+        if (auto *RHS = dyn_cast<ConstantInt>(ICmp->getOperand(1))) {
```

Comment:
Rather than implementing your own limited constant folding here, it would be better to call ConstantFoldCompareInstOperands.

Reply:
I was only trying to deal with the pattern that SimplifyCFG creates for itself - not general constant folding. But, point taken...
Anyway - need to decide if I leave it here or this logic needs to move to where PHIs are simplified...